pkg/oauth/external/handler.go

```diff
@@ -180,29 +176,48 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	}
 
 	klog.V(5).Infof("Got access data")
+	h.login(w, req, accessData, authData.State)
+}
 
-	identity, ok, err := h.provider.GetUserIdentity(accessData)
+func (h *Handler) login(w http.ResponseWriter, req *http.Request, accessData *osincli.AccessData, state string) {
+	identity, err := h.provider.GetUserIdentity(accessData)
 	if err != nil {
+		var authorizationDeniedError AuthorizationDeniedError
+		if errors.As(err, &authorizationDeniedError) {
+			klog.V(4).Infof("Authorization denied: %v", authorizationDeniedError)
+			audit.AddUsernameAnnotation(req, authorizationDeniedError.Identity().GetProviderPreferredUserName())
+			audit.AddDecisionAnnotation(req, audit.DenyDecision)
+			h.handleError(err, w, req)
+			return
+		}
+
+		var authorizationFailedError AuthorizationFailedError
+		if errors.As(err, &authorizationFailedError) {
+			klog.V(4).Infof("Authorization failed: %v", authorizationFailedError)
+			audit.AddUsernameAnnotation(req, authorizationFailedError.Identity().GetProviderPreferredUserName())
+			audit.AddDecisionAnnotation(req, audit.ErrorDecision)
+			h.handleError(err, w, req)
+			return
+		}
+
 		klog.V(4).Infof("Error getting userIdentityInfo info: %v", err)
-		h.handleError(err, w, req)
-		return
-	}
-	if !ok {
-		klog.V(4).Infof("Could not get userIdentityInfo info from access token")
-		err := errors.New("Could not get userIdentityInfo info from access token")
+		audit.AddDecisionAnnotation(req, audit.ErrorDecision)
 		h.handleError(err, w, req)
```

Comment:
I think switch case would be more readable here:
```go
var authorizationDeniedError AuthorizationDeniedError
var authorizationFailedError AuthorizationFailedError
switch {
    case errors.As(err, &authorizationDeniedError):
         ....
    case errors.As(err, &authorizationFailedError):
        ...
    default:
        ...
}
```
I'll leave that up to you though

Reply:
done